rfcs/0003-declaration-engine.md

```diff
@@ -0,0 +1,189 @@
+
+### Definition
+
+During type checking of a declaration, the compiler will add information about that declaration to the declaration engine:
+
+```rust
+struct DeclarationEngine {
+    // function name to typed function declaration
+    // this example code does not consider Path's, but we could use
+    // a BTreeMap with Path's as the key
+    functions: HashMap<String, TypedFunctionDeclaration>,
+
+    // struct name to typed struct declaration
+    // again, this does not consider Path's
+    structs: HashMap<String, TypedStructDeclaration>,
+
+    // enums and traits
+}
+
+impl DeclarationEngine {
+    fn add_function(&mut self, function: TypedFunctionDeclaration) -> DeclarationDefinition {
+        // add to self and return DeclarationDefinition
+    }
+
+    fn add_struct(&mut self, r#struct: TypedStructDeclaration) -> DeclarationDefinition {
+        // add to self and return DeclarationDefinition
+    },
+
+    // enums and traits
+}
+
+enum DeclarationDefinition {
+    // does not consider Path's
+    Function(String),
+
+    // does not consider Path's
+    Struct(String),
+
+    // enums and traits
+}
+```
+
+Then in the typed AST, that declarations could look like (in pseudocode):
+
+```rust
+contract;
+
+/// imports
+
+DeclarationDefinition::Function(<function name>)
+DeclarationDefinition::Struct(<struct name>)
+
+/// storage declaration
+
+DeclarationDefinition::Function("main")
+
+```
+
+Because inlining is still needed for the IR, these instances of `DeclarationDefinition` would be replaced in an additional step at the end of type checking called the [resolution step](https://github.com/FuelLabs/sway/issues/1820).
+
+### Usage
+
+When a declaration is _used_ an internal reference is created that references the existence of a possible declaration in the declaration engine, and that is resolved during the resolution step. Given:
+
+```rust
+fn add(x: u64, y: u64) -> u64 {
+    ..
+}
+```
+
+The `TypedFunctionDeclaration` for `add` is added to the declaration engine. Then given:
+
+```rust
+let foo = add(1, 2);
+```
+
+`foo` will be a `TypedExpression` with a [`TypedExpressionVariant::FunctionApplication`](https://github.com/FuelLabs/sway/blob/dcaee960df9b3466c6897d0a86ea806c9abf7edd/sway-core/src/semantic_analysis/ast_node/expression/typed_expression_variant.rs#L18), where instead of storing the entire `TypedFunctionDeclaration` in the `TypedExpressionVariant::FunctionApplication` variant, there is a reference created `Function("add", vec!(u64, u64))` which is stored [here](https://github.com/FuelLabs/sway/blob/dcaee960df9b3466c6897d0a86ea806c9abf7edd/sway-core/src/semantic_analysis/ast_node/expression/typed_expression_variant.rs#L23). The type of `foo` is some new `TypeInfo::ReturnType("add", vec!(u64, u64))`. Both the reference `Function("add", vec!(u64, u64))` and `TypeInfo::ReturnType("add", vec!(u64, u64))` are resolved during the resolution step.
```

Comment:
```suggestion
`foo` will be a `TypedExpression` with a [`TypedExpressionVariant::FunctionApplication`](https://github.com/FuelLabs/sway/blob/dcaee960df9b3466c6897d0a86ea806c9abf7edd/sway-core/src/semantic_analysis/ast_node/expression/typed_expression_variant.rs#L18), where instead of storing the entire `TypedFunctionDeclaration` in the `TypedExpressionVariant::FunctionApplication` variant, there is a reference created `Function("add", vec![u64, u64])` which is stored [here](https://github.com/FuelLabs/sway/blob/dcaee960df9b3466c6897d0a86ea806c9abf7edd/sway-core/src/semantic_analysis/ast_node/expression/typed_expression_variant.rs#L23). The type of `foo` is some new `TypeInfo::ReturnType("add", vec![u64, u64])`. Both the reference `Function("add", vec![u64, u64])` and `TypeInfo::ReturnType("add", vec![u64, u64])` are resolved during the resolution step.
```